### PR TITLE
Update devcontainer to slim image and other enhancements

### PR DESCRIPTION
**Pull request by usrbinkat**
## Description

This PR represents enhancements based on feedback and usage of the devcontainer.

Some of the changes include:

* Add `Open in Github Codespaces` button to readme
* Adopt the new flavors of devcontainer
* Default the devcontainer to `slim-hugo` which is much smaller
* Cleanup directory layout to rm messy .devcontainer* repo root artifacts
* Enhance Makefile to auto-lint correct devcontainer.json files
* Improved container internals for reliability
* Improved devcontainer.json for better vscode configuration and use
* Manually confirmed README make assess and make lint work as expected

## Checklist:

- [x] I have reviewed the [style guide](https://github.com/pulumi/pulumi-hugo/blob/master/STYLE-GUIDE.md).
- [ ] If blogging, I have reviewed the [blogging guide](https://github.com/pulumi/pulumi-hugo/blob/master/BLOGGING.md).
- [x] I have manually confirmed that all new links work.
- [ ] I added aliases (i.e., redirects) for all filename changes.
- [ ] If making css changes, I rebuilt the bundle.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # pulumi-hugo
 
+[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/pulumi/pulumi-hugo)
```
Comment (cnunciato) [inline README.md:3]: Any reason not to put this under the VS Code devcontainer heading? (I'm not sure what the conventions are around this, but assuming this launches/uses the devcontainer, it'd seem to make sense to put it there.)

File: .gitmodules [hunk 2]
```diff
@@ -1,3 +1,3 @@
-[submodule ".devcontainer"]
-	path = .devcontainer
+[submodule ".github/devcontainer"]
```
Comment (cnunciato) [inline .gitmodules:1]: Somehow makes me _slightly_ nervous to use `.github` for this, given how heavily rely on GHA workflows. Is the convention to use this dir for this, and with submodule?

**Review by cnunciato — COMMENTED** [review]
LGTM, just a couple questions though.

File: README.md (hunk 1 above)
Comment (usrbinkat) [reply README.md:3]: I can relocate it there if you prefer. I was intentional in setting it at the top for convenient and obvious "oh i just do this to get started" but of course it is reasonable to nest it in the appropriate section.

File: .gitmodules (hunk 2 above)
Comment (usrbinkat) [reply .gitmodules:1]: .github dir is a general purpose directory. all workflows files should be isolated in the .github/workflows directory so .github/devcontainer is a non-conflicting path.

I can research and add some documentation as I find it, I'll include opinions/conventions documented both for and against this type of use case expansion.

Let me do that real quick.

File: Makefile [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
-	cp .devcontainer/devcontainer.json .devcontainer.json
+devcontainer: ensure
+	mkdir -p .devcontainer
+	git submodule update --init .github/devcontainer
+	git submodule update --remote --merge .github/devcontainer
+	rsync -av .github/devcontainer/.devcontainer .
+	yarn prettier --write .devcontainer/*.json
+	yarn prettier --write .devcontainer/*/*.json
+	cp -f .devcontainer/hugo/* .devcontainer/
+	rm -rf .github/devcontainer/* .github/devcontainer/.* 2>/dev/null || true
```
Comment (cnunciato) [inline Makefile:12]:
```suggestion
devcontainer:
	mkdir -p .devcontainer
	git submodule update --init .github/devcontainer
	git submodule update --remote --merge .github/devcontainer
	rsync -av .github/devcontainer/.devcontainer .
	cp -f .devcontainer/hugo/* .devcontainer/
	rm -rf .github/devcontainer/* .github/devcontainer/.* 2>/dev/null || true
```

File: Makefile [hunk 4]
```diff
@@ -16,7 +20,7 @@ ensure:
 	./scripts/ensure.sh
 
 .PHONY: lint
-lint:
+lint: ensure
```
Comment (cnunciato) [inline Makefile:23]:
```suggestion
lint:
```

**Review by cnunciato — COMMENTED** [review]
We chatted in DM, but for the PR, we can remove these deps on `ensure` and just ignore Prettifiying the devcontainer JSON files.

File: README.md (hunk 1 above)
Comment (usrbinkat) [reply README.md]: Resolved in [26e76ba](https://github.com/pulumi/pulumi-hugo/pull/3909/commits/26e76ba2785d17f0ff11942a25a01ee660ec2790)

File: .gitmodules (hunk 2 above)
Comment (usrbinkat) [reply .gitmodules:1]: Here are my references.

* Example of expanded .github usage & hygiene: 
  * [github.com/microsoft/terminal/.github](https://github.com/microsoft/terminal/tree/main/.github)
  * [docs.github.com/en/actions/learn-github-actions](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions)
  * [opensource.creativecommons.org/contributing-code/github-repo-guidelines](https://opensource.creativecommons.org/contributing-code/github-repo-guidelines/)
  * [probot.github.io/docs/best-practices](https://probot.github.io/docs/best-practices/#include-dry-run-functionality)
  * [scroll to .github/scripts definition](https://aws.amazon.com/blogs/devops/integrating-with-github-actions-ci-cd-pipeline-to-deploy-a-web-app-to-amazon-ec2/)
  * [github.com/joelparkerhenderson/github-special-files-and-path](https://github.com/joelparkerhenderson/github-special-files-and-paths#introduction)
  * [.github/actions](wallis.dev/blog/composite-github-actions)

The only well defined argument alternative I've found so far is [dot-config.github.io](https://dot-config.github.io/) which argues in favor of a standard root .config directory. There is more conversation on the topic on [ycombinator.com](https://news.ycombinator.com/item?id=24066748) of course.

Comment (usrbinkat) [reply .gitmodules:1]: The general conversation seems to indicate the purpose of .github dir being for any non-conflicting use-case that supports the regular maintenance and operation of the repository's github related integrations.

For this reason I believe .github/devcontainer is a reasonable github asset which upon careful research does not conflict with any other path by that name.

Comment (cnunciato) [reply .gitmodules:1]: Good enough for me, and seems like a thing we can always change if we need to. Thanks for looking into it!

**Review by cnunciato — APPROVED** [review]

**Pull request merged**